SimPPS/PPSPixelDigiProducer/src/RPixLinearChargeDivider.cc

```diff
@@ -0,0 +1,107 @@
+                                                 uint32_t det_id)
+    : rndEngine_(eng), det_id_(det_id) {
+  verbosity_ = params.getParameter<int>("RPixVerbosity");
+  fluctuate = new SiG4UniversalFluctuation();
+  fluctuateCharge_ = params.getParameter<bool>("RPixLandauFluctuations");
+  chargedivisions_ = params.getParameter<int>("RPixChargeDivisions");
+  deltaCut_ = params.getParameter<double>("RPixDeltaProductionCut");
+}
+
+RPixLinearChargeDivider::~RPixLinearChargeDivider() { delete fluctuate; }
+
+std::vector<RPixEnergyDepositUnit> RPixLinearChargeDivider::divide(const PSimHit& hit) {
+  LocalVector direction = hit.exitPoint() - hit.entryPoint();
+  if (direction.z() > 10 || direction.x() > 200 || direction.y() > 200) {
+    the_energy_path_distribution_.clear();
+    return the_energy_path_distribution_;
+  }
+
+  int NumberOfSegmentation = chargedivisions_;
+  double eLoss = hit.energyLoss();  // Eloss in GeV
+  the_energy_path_distribution_.resize(NumberOfSegmentation);
+
+  if (fluctuateCharge_) {
+    int pid = hit.particleType();
+    double momentum = hit.pabs();
+    double length = direction.mag();  // Track length in Silicon
+    FluctuateEloss(pid, momentum, eLoss, length, NumberOfSegmentation, the_energy_path_distribution_);
+    for (int i = 0; i < NumberOfSegmentation; i++) {
+      the_energy_path_distribution_[i].Position() =
+          hit.entryPoint() + double((i + 0.5) / NumberOfSegmentation) * direction;
+    }
+  } else {
+    for (int i = 0; i < NumberOfSegmentation; i++) {
+      the_energy_path_distribution_[i].Position() =
+          hit.entryPoint() + double((i + 0.5) / NumberOfSegmentation) * direction;
+      the_energy_path_distribution_[i].Energy() = eLoss / (double)NumberOfSegmentation;
+    }
+  }
+
+  if (verbosity_) {
+    edm::LogInfo("RPixLinearChargeDivider") << det_id_ << " charge along the track:";
+    double sum = 0;
+    for (unsigned int i = 0; i < the_energy_path_distribution_.size(); i++) {
+      edm::LogInfo("RPixLinearChargeDivider")
+          << the_energy_path_distribution_[i].X() << " " << the_energy_path_distribution_[i].Y() << " "
+          << the_energy_path_distribution_[i].Z() << " " << the_energy_path_distribution_[i].Energy();
+      sum += the_energy_path_distribution_[i].Energy();
+    }
+    edm::LogInfo("RPixLinearChargeDivider") << "energy dep. sum=" << sum;
+  }
+
+  return the_energy_path_distribution_;
+}
+
+void RPixLinearChargeDivider::FluctuateEloss(int pid,
+                                             double particleMomentum,
+                                             double eloss,
+                                             double length,
+                                             int NumberOfSegs,
+                                             std::vector<RPixEnergyDepositUnit>& elossVector) {
+  double particleMass = 139.6;  // Mass in MeV, Assume pion
+  pid = std::abs(pid);
+  if (pid != 211) {  // Mass in MeV
```

Comment:
@mundim , in PPS we handle only protons, logic of pid and particleMass should correspond for proton only.

Reply:
In fact, due to interaction with the material, other particles might occur. Debugging the code it was found many other kinds of particles present.